internal/hashing/argon2id.go

```diff
@@ -0,0 +1,162 @@
+package hashing
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"golang.org/x/crypto/argon2"
+)
+
+const ARGON2_VERSION = 19
+
+type Argon2IDParameters struct {
+	SaltLength uint32
+	Time       uint32
+	Memory     uint32
+	Threads    uint8
+	KeyLength  uint32
+}
+
+// Default parameters recommended by OWASP
+var DefaultArgon2IDParameters = Argon2IDParameters{
+	SaltLength: 16,
+	Time:       2,
+	Memory:     19 * 1024, // 19 MiB
+	Threads:    1,
+	KeyLength:  32,
+}
+
+var ErrInvalidPassword = errors.New("invalid password")
+
+// Takes the password string, makes sure it's not empty, and hashes it using argon2id
+func HashPassword(password string) (string, error) {
+	if strings.TrimSpace(password) == "" {
+		return "", fmt.Errorf("%w: password cannot be empty", ErrInvalidPassword)
+	}
+	salt, err := generateSalt(DefaultArgon2IDParameters.SaltLength)
+	if err != nil {
+		return "", err
+	}
+	hash := hashPasswordWithSaltAndParams(password, salt, DefaultArgon2IDParameters)
+
+	encoded_salt := base64.RawStdEncoding.EncodeToString(salt)
+	encoded_hash := base64.RawStdEncoding.EncodeToString(hash)
+	hashedPassword := fmt.Sprintf("$argon2id$v=%d$m=%d,t=%d,p=%d$%s$%s",
```

Comment:
I think it would be simpler to have an encode and decode function for the hashed password.

```
func encode(params Argon2IDParameters) (string, error)
func decode(encoded string) (*Argon2IDParameters, error) 
```

They seem to do something similar [here](https://go-review.googlesource.com/c/crypto/+/431595/1/argon2/encoding.go#48)

Reply:
I now use a similar approach. I kept the public struct `Argon2IDParameters` separate, as this will eventually be used to configure the hashing parameters. Internally, I added the `phcEncoding` struct.